display/SINK.tally/tally_callbacks.cpp

```diff
@@ -0,0 +1,205 @@
+static void lttng_host_usr_callback(
+    struct common_data_s *common_data, void *usr_data,   const char* hostname,
+    int64_t vpid, uint64_t vtid, int64_t ts, int64_t backend_id, const char* name,
+    uint64_t dur, bt_bool err
+) 
+{
+    /* In callbacks, the user just  need to cast our API usr_data to his/her data structure */
+    struct tally_dispatch *data = (struct tally_dispatch *) usr_data;
+
+    TallyCoreTime a{dur, err};
+    const int level = backend_level[backend_id];
+    data->host_backend_name[level].insert(backend_name[backend_id]);
+    data->host[level][hpt_function_name_t(hostname, vpid, vtid, name)] += a;
+}
+
+static void lttng_device_usr_callback(
+    common_data_t *common_data, void *usr_data, const char* hostname, int64_t vpid,
+    uint64_t vtid, int64_t ts, int64_t backend, const char* name, uint64_t dur, 
+    uint64_t did, uint64_t sdid, bt_bool err, const char* metadata
+)
+{
+    /* In callbacks, the user just  need to cast our API usr_data to his/her data structure */
+    struct tally_dispatch *data = (struct tally_dispatch *) usr_data;
+
+    /* TODO: Should fucking cache this function */
+    const auto name_demangled = (data->demangle_name) ? f_demangle_name(name) : name;
+    const auto name_with_metadata = (data->display_kernel_verbose && !strcmp(metadata, "")) ? name_demangled + "[" + metadata + "]" : name_demangled;
+
+    TallyCoreTime a{dur, err};
+    data->device[hpt_device_function_name_t(hostname, vpid, vtid, did, sdid, name_with_metadata)] += a;
+
+}
+
+static void lttng_traffic_usr_callback(
+    common_data_t *common_data, void *usr_data, const char* hostname,
+    int64_t vpid, uint64_t vtid, int64_t ts, int64_t backend, 
+    const char* name, uint64_t size
+
+)
+{
+    /* In callbacks, the user just  need to cast our API usr_data to his/her data structure */
+    struct tally_dispatch *data = (struct tally_dispatch *) usr_data;
+
+    TallyCoreByte a{(uint64_t)size, false};
+    const int level = backend_level[backend];
+    data->traffic_backend_name[level].insert(backend_name[backend]);
+    data->traffic[level][hpt_function_name_t(hostname, vpid, vtid, name)] += a;
+}
+
+static void lttng_device_name_usr_callback(
+    common_data_t *common_data, void *usr_data, const char* hostname, int64_t vpid,
+    uint64_t vtid, int64_t ts, int64_t backend, const char* name, uint64_t did
+)
+{
+    /* In callbacks, the user just  need to cast our API usr_data to his/her data structure */
+    struct tally_dispatch *data = (struct tally_dispatch *) usr_data;
+
+    data->device_name[hp_device_t(hostname, vpid, did)] = name;
+}
+
+static void lttng_thapi_metadata_usr_callback(
+    common_data_t *common_data, void *usr_data,const char* hostname, int64_t vpid,
+    uint64_t vtid, int64_t ts, int64_t backend, const char* metadata
+)
+{
+    /* In callbacks, the user just  need to cast our API usr_data to his/her data structure */
+    struct tally_dispatch *data = (struct tally_dispatch *) usr_data;
+
```

Comment:
If metadata is only a `const char*` in input why `data->metadat`a a vector? Maybe a bug...

Reply:
I think the implementation is the right one, the user can use lttng_thapi:metadata several times in the code, so here your accumulate these messages in a std::vector<const * char>.

Reply:
Oh good call!